File: backend-py/app/apis/booking.py
import random
import string
from flask import request, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from flask_restx import Namespace, Resource, fields, marshal, reqparse
from marshmallow import ValidationError
from sqlalchemy import text
from sqlalchemy.orm import sessionmaker

from app.apis.flight import flight_model_output
from app.apis.utils import price_from_flight
from app.core.auth import roles_required
from app.extensions import db
from app.models.booking import (
    Booking,
    BookingDepartureFlight,
    BookingReturnFlight,
    BookingFlightExtra,
)
from app.models.common import ClassType
from app.models.flight import Flight, FlightExtra
from app.models.seat_session import SeatSession
from app.schemas.booking import (
    booking_schema,
    booking_output_schema,
    bookings_output_schema,
)

api = Namespace("booking", description="Booking related operations")

# --- RESTx Models ---

extra_model_input = api.model(
    "ExtraInput",
    {
        "id": fields.String(required=True, description="Extra ID"),
        "quantity": fields.Integer(required=True, description="Quantity"),
    },
)

booking_model_input = api.model(
    "BookingInput",
    {
        "session_id": fields.String(required=True, description="Session ID"),
        "departure_flights": fields.List(
            fields.String, required=True, description="Departure flights ID"
        ),
        "return_flights": fields.List(
            fields.String, required=True, description="Return flights ID"
        ),
        "extras": fields.List(
            fields.Nested(extra_model_input), required=True, description="List of extras selected"
        ),
        "has_booking_insurance": fields.Boolean(
            required=True, description="Whether insurance was purchased"
        ),
    },
)

booked_flight_extra_model_output = api.model(
    "BookedFlightExtraOutput",
    {
        "extra_id": fields.String(readonly=True, description="FlightExtra ID"),
        "extra_price": fields.Float(required=True, description="Price"),
        "name": fields.String(required=True, description="Name"),
        "description": fields.String(required=True, description="Description"),
        "quantity": fields.Integer(required=True, description="Quantity"),
    },
)

booked_flight_model_output = api.model(
    "BookedFlightOutput",
    {
        "flight": fields.Nested(flight_model_output, description="Flight"),
        "seat_number": fields.String(required=True, description="Seat number"),
        "class_type": fields.String(
            enum=[e.name for e in ClassType], required=True, description="Class type"
        ),
        "price": fields.Float(required=True, description="Price"),
        "extras": fields.List(
            fields.Nested(booked_flight_extra_model_output),
            description="List of extra service IDs",
        ),
    },
)

booking_model_output = api.model(
    "BookingOutput",
    {
        "id": fields.String(readonly=True, description="Booking ID"),
        "booking_number": fields.String(readonly=True, description="Booking number"),
        "departure_flights": fields.List(
            fields.Nested(booked_flight_model_output), description="Departure flights"
        ),
        "return_flights": fields.List(
            fields.Nested(booked_flight_model_output), description="Return flights"
        ),
        "total_price": fields.Float(required=True, description="Total price"),
        "is_insurance_purchased": fields.Boolean(
            description="Whether insurance was purchased"
        ),
        "insurance_price": fields.Float(required=True, description="Insurance price"),
    },
)

# --- Request Parsers ---
booking_list_parser = reqparse.RequestParser()
booking_list_parser.add_argument(
    "flight_id", type=str, help="Filter by flight ID", location="args"
)
booking_list_parser.add_argument(
    "user_id", type=str, help="Filter by user ID", location="args"
)
booking_list_parser.add_argument(
    "class_type", type=str, help="Filter by class type", location="args"
)

def generate_unique_booking_number(session):
    while True:
        candidate = "".join(random.choices(string.ascii_uppercase, k=6))
        exists = session.query(Booking).filter_by(booking_number=candidate).first()
        if not exists:
            return candidate
        
def check_and_update_flight_capacity(sql_session, flight_id):
    """Check if a flight is fully booked and update the flag"""
    flight = sql_session.query(Flight).filter(Flight.id == flight_id).first()
    if not flight:
        return
    
    # Get total seats available on the aircraft
    total_seats = (
        len(flight.aircraft.first_class_seats) +
        len(flight.aircraft.business_class_seats) +
        len(flight.aircraft.economy_class_seats)
    )
    
    # Get total booked seats for this flight
    booked_seats_count = len(flight.booked_seats_confirmed)
    
    # Update fully_booked flag
    flight.fully_booked = booked_seats_count >= total_seats
    sql_session.commit()

@api.route("/")
@api.response(500, "Internal Server Error")
class BookingList(Resource):
    @jwt_required()
    @api.expect(booking_list_parser)
    @api.response(200, "OK", [booking_model_output])
    @api.response(403, "Forbidden")
    @api.response(404, "Not Found")
    @api.response(400, "Bad Request")
    def get(self):
        """List bookings with optional filtering"""
        user_id = get_jwt_identity()
        args = booking_list_parser.parse_args()

        # Regular users can only see their own bookings
        datastore = current_app.extensions["security"].datastore
        user = datastore.find_user(id=user_id)

        query = Booking.query

        if not user.has_role("admin") and not user.has_role("airline-admin"):
            # Regular users can only see their own bookings
            query = query.filter(Booking.user_id == user_id)
        elif user.has_role("airline-admin"):
            if not user.airline_id:
                return {"error": "Airline ID is required for airline-admin"}, 400
            # Airline admins can see bookings for their flights
            query = query.join(Flight).filter(Flight.airline_id == user.airline_id)

        # Apply filters
        if args["flight_id"]:
            query = query.filter(Booking.flight_id == args["flight_id"])
        if args["class_type"]:
            query = query.filter(Booking.class_type == args["class_type"])
        if args["user_id"] and (
                user.has_role("admin") or user.has_role("airline-admin")
        ):
            query = query.filter(Booking.user_id == args["user_id"])

        return (
            marshal(bookings_output_schema.dump(query.all()), booking_model_output),
            200,
        )


    @api.expect(booking_model_input)
    @jwt_required()
    @roles_required(["user"])
    @api.response(201, "Created", api.model("BookingCreationResponse", {
        "id": fields.String(readonly=True, description="Booking ID")
    }))
    @api.response(400, "Bad Request")
    @api.response(404, "Not Found")
    def post(self):
        """Create a new booking"""
        # TODO: AGGIUNGERE LOCK TRANZAZIONALE CHE NON VENGA ELIMINATA LA SESSIONEE
        # CHECK SESSION HIJACKING

        user_id = get_jwt_identity()
        data = request.json

        with db.engine.begin() as connection:
            connection.execute(text("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE"))

            # Use a new session bound to this connection
            session = sessionmaker(bind=connection)
            sql_session = session()

            try:
                validated_data = booking_schema.load(data)
            except ValidationError as err:
                return {"errors": err.messages}, 400

            seat_session = (
                sql_session.query(SeatSession)
                .filter(
                    SeatSession.id == validated_data["session_id"],
                    SeatSession.user_id == user_id,
                )
                .first()
            )
            if seat_session is None:
                return {
                    "error": "Seat session does not belong to the user",
                    "code": 403,
                }, 403
            all_seats = seat_session.seats
            booking = Booking(
                user_id=user_id,
                payment_confirmed=True,
                has_booking_insurance=validated_data["has_booking_insurance"],
                booking_number=generate_unique_booking_number(sql_session),
            )
            sql_session.add(booking)
            sql_session.commit()

            departure_flights = validated_data["departure_flights"]
            return_flights = validated_data["return_flights"]
            
            extras = validated_data["extras"]

            for flight_id in departure_flights:
                for seat in all_seats:
                    if seat.flight_id == flight_id:
                        flight = (
                            sql_session.query(Flight)
                            .filter(Flight.id == flight_id)
                            .first()
                        )
                        class_type = seat.class_type
                        flight_price = price_from_flight(flight, class_type)
                        booking_flight = BookingDepartureFlight(
                            flight_id=flight_id,
                            booking_id=booking.id,
                            seat_number=seat.seat_number,
                            class_type=class_type,
                            price=flight_price,
                        )
                        sql_session.add(booking_flight)
                        sql_session.flush()
                        check_and_update_flight_capacity(sql_session, flight_id)
                        
                        
                        

            for flight_id in return_flights:
                for seat in all_seats:
                    if seat.flight_id == flight_id:
                        flight = (
                            sql_session.query(Flight)
                            .filter(Flight.id == flight_id)
                            .first()
                        )
                        class_type = seat.class_type
                        flight_price = price_from_flight(flight, class_type)
                        booking_flight = BookingReturnFlight(
                            flight_id=flight_id,
                            booking_id=booking.id,
                            seat_number=seat.seat_number,
                            class_type=class_type,
                            price=flight_price,
                        )
                        sql_session.add(booking_flight)
                        sql_session.flush()
                        check_and_update_flight_capacity(sql_session, flight_id)

            for extra in extras:
                extra_id = extra["id"]
                quantity = extra["quantity"]
                extra_obj = (
                    sql_session.query(FlightExtra)
                    .filter(FlightExtra.id == extra_id)
                    .first()
                )
                if extra_obj:
                    if extra_obj.flight_id not in departure_flights and extra_obj.flight_id not in return_flights:
                        return {
                            "error": "Extra does not belong to the selected flights",
                            "code": 403,
                        }, 403
                        
                    booking_flight_extra = BookingFlightExtra(
                        booking_id=booking.id,
                        flight_id=extra_obj.flight_id,
                        extra_id=extra_obj.id,
                        extra_price=extra_obj.price * quantity,
                        quantity=quantity,
                    )
                    sql_session.add(booking_flight_extra)
            sql_session.commit()
            sql_session.delete(seat_session)
            sql_session.commit()

            return {"id": str(booking.id)}, 201


@api.route("/<uuid:booking_id>")
@api.param("booking_id", "The booking identifier")
class BookingResource(Resource):
    @jwt_required()
    @api.response(200, "OK", booking_model_output)
    @api.response(403, "Forbidden")
    @api.response(404, "Not Found")
    def get(self, booking_id):
        """Fetch a booking given its identifier"""
        user_id = get_jwt_identity()
        datastore = current_app.extensions["security"].datastore
        user = datastore.find_user(id=user_id)

        booking = Booking.query.get_or_404(booking_id)

        # Check permissions
        if (
                str(booking.user_id) != user_id
                and not user.has_role("admin")
                and not (user.has_role("airline-admin") and user.airline_id)
        ):
            return {"error": "You do not have permission to view this booking"}, 403

        # For airline admins, check if booking is for their airline's flight
        if user.has_role("airline-admin") and user.airline_id:
            flight = Flight.query.get(booking.flight_id)
            if flight.airline_id != user.airline_id:
                return {"error": "You do not have permission to view this booking"}, 403

        return marshal(booking_output_schema.dump(booking), booking_model_output), 200

    @jwt_required()
    @api.response(200, "OK")
    @api.response(403, "Forbidden")
    @api.response(404, "Not Found")
    def delete(self, booking_id):
        """Delete a booking given its identifier"""
        user_id = get_jwt_identity()
        booking = Booking.query.get_or_404(booking_id)

        # Check permissions
        if str(booking.user_id) != user_id:
            datastore = current_app.extensions["security"].datastore
            user = datastore.find_user(id=user_id)
            if not user.has_role("admin"):
                return {
                    "error": "You do not have permission to delete this booking",
                    "code": 403,
                }, 403
                
                
                # Collect flight IDs before deletion to update capacity
        flight_ids = set()
        for departure_flight in booking.departure_bookings:
            flight_ids.add(departure_flight.flight_id)
        for return_flight in booking.return_bookings:
            flight_ids.add(return_flight.flight_id)

        db.session.delete(booking)
        db.session.commit()

        # Update flight capacities
        for flight_id in flight_ids:
            check_and_update_flight_capacity(db.session, flight_id)
        
        db.session.commit()
        return {"message": "Booking deleted successfully"}, 200

File: backend-ts/src/apis/booking.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { randomUUID } from 'crypto';
import { PrismaClient } from '../../generated/prisma';
import { authenticateToken, requireRoles } from '../middleware/auth';
import { registry } from '../config/openapi';
import { 
  bookingInputSchema, 
  bookingListQuerySchema,
  bookingOutputSchema,
  BookingInput,
  BookingListQuery 
} from '../schemas/booking';
import { 
  priceFromFlight, 
  generateUniqueBookingNumber, 
  checkAndUpdateFlightCapacity 
} from '../utils/booking';

const router = Router();
const prisma = new PrismaClient();

// Validation middleware
const validateBookingInput = (req: Request, res: Response, next: any) => {
  try {
    // Validate request body without reassigning
    bookingInputSchema.parse(req.body);
    next();
  } catch (error) {
    if (error instanceof z.ZodError) {
      res.status(400).json({ errors: error.errors });
      return;
    }
    next(error);
  }
};

const validateBookingListQuery = (req: Request, res: Response, next: any) => {
  try {
    // Validate query parameters without reassigning
    bookingListQuerySchema.parse(req.query);
    next();
  } catch (error) {
    if (error instanceof z.ZodError) {
      res.status(400).json({ errors: error.errors });
      return;
    }
    next(error);
  }
};

// Helper function to format booking output (matches Python schemas)
async function formatBookingOutput(booking: any) {
  // Helper function to format flight data to match Python structure
  const formatFlightData = (flightData: any) => {
    const flight = flightData.flight;
    return {
      ...flight,
      airline: flight.route.airline,
      flight_number: flight.route.flight_number,
      departure_airport: flight.route.airport_route_departure_airport_idToairport,
      arrival_airport: flight.route.airport_route_arrival_airport_idToairport,
      aircraft: flight.airline_aircraft
    };
  };

  // Calculate insurance price (matches Python model property)
  const insurancePrice = Math.round(
    (booking.booking_departure_flight.reduce((sum: number, flight: any) => 
      sum + flight.flight.price_insurance, 0) +
    booking.booking_return_flight.reduce((sum: number, flight: any) => 
      sum + flight.flight.price_insurance, 0)) * 100
  ) / 100;

  // Calculate total price (matches Python model property)
  let totalPrice = 0;
  
  // Add departure flights price and extras
  for (const flight of booking.booking_departure_flight) {
    totalPrice += flight.price;
  }
  
  // Add return flights price and extras
  for (const flight of booking.booking_return_flight) {
    totalPrice += flight.price;
  }
  
  // Add booking extras
  for (const extra of booking.booking_flight_extra || []) {
    totalPrice += extra.extra_price;
  }
  
  // Add insurance if purchased
  if (booking.has_booking_insurance) {
    totalPrice += insurancePrice;
  }
  
  totalPrice = Math.round(totalPrice * 100) / 100;

  // Group extras by flight
  const extrasByFlight = new Map();
  for (const extra of booking.booking_flight_extra || []) {
    if (!extrasByFlight.has(extra.flight_id)) {
      extrasByFlight.set(extra.flight_id, []);
    }
    extrasByFlight.get(extra.flight_id).push({
      extra_id: extra.extra_id,
      extra_price: extra.extra_price,
      name: extra.flight_extra.extra.name,
      description: extra.flight_extra.extra.description,
      quantity: extra.quantity
    });
  }

  return {
    id: booking.id,
    booking_number: booking.booking_number,
    departure_flights: booking.booking_departure_flight.map((flight: any) => ({
      flight: formatFlightData(flight),
      seat_number: flight.seat_number,
      class_type: flight.class_type,
      price: flight.price,
      extras: extrasByFlight.get(flight.flight_id) || []
    })),
    return_flights: booking.booking_return_flight.map((flight: any) => ({
      flight: formatFlightData(flight),
      seat_number: flight.seat_number,
      class_type: flight.class_type,
      price: flight.price,
      extras: extrasByFlight.get(flight.flight_id) || []
    })),
    total_price: totalPrice,
    is_insurance_purchased: booking.has_booking_insurance,
    insurance_price: insurancePrice
  };
}


router.get('/', authenticateToken, validateBookingListQuery, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.id;
    const userRoles = req.user!.roles;
    const query = req.query as BookingListQuery;

    // Build where clause based on user permissions (matches Python logic)
    let whereClause: any = {};

    const isAdmin = userRoles.includes('admin');
    const isAirlineAdmin = userRoles.includes('airline-admin');

    if (!isAdmin && !isAirlineAdmin) {
      // Regular users can only see their own bookings
      whereClause.user_id = userId;
    } else if (isAirlineAdmin) {
      // Get user's airline_id
      const user = await prisma.user.findUnique({
        where: { id: userId },
        select: { airline_id: true }
      });

      if (!user?.airline_id) {
        res.status(400).json({ error: 'Airline ID is required for airline-admin' });
        return;
      }

      // Airline admins can see bookings for their flights
      whereClause = {
        OR: [
          {
            booking_departure_flight: {
              some: {
                flight: {
                  route: {
                    airline_id: user.airline_id
                  }
                }
              }
            }
          },
          {
            booking_return_flight: {
              some: {
                flight: {
                  route: {
                    airline_id: user.airline_id
                  }
                }
              }
            }
          }
        ]
      };
    }

    // Apply additional filters
    if (query.flight_id) {
      whereClause.OR = [
        {
          booking_departure_flight: {
            some: { flight_id: query.flight_id }
          }
        },
        {
          booking_return_flight: {
            some: { flight_id: query.flight_id }
          }
        }
      ];
    }

    if (query.class_type) {
      whereClause.OR = [
        {
          booking_departure_flight: {
            some: { class_type: query.class_type }
          }
        },
        {
          booking_return_flight: {
            some: { class_type: query.class_type }
          }
        }
      ];
    }

    if (query.user_id && (isAdmin || isAirlineAdmin)) {
      whereClause.user_id = query.user_id;
    }

    const bookings = await prisma.booking.findMany({
      where: whereClause,
      include: {
        booking_departure_flight: {
          include: {
            flight: {
              include: {
                route: {
                  include: {
                    airport_route_departure_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airport_route_arrival_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airline: true
                  }
                },
                airline_aircraft: {
                  include: {
                    aircraft: true
                  }
                }
              }
            }
          }
        },
        booking_return_flight: {
          include: {
            flight: {
              include: {
                route: {
                  include: {
                    airport_route_departure_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airport_route_arrival_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airline: true
                  }
                },
                airline_aircraft: {
                  include: {
                    aircraft: true
                  }
                }
              }
            }
          }
        },
        booking_flight_extra: {
          include: {
            flight_extra: {
              include: {
                extra: true
              }
            }
          }
        }
      }
    });

    const formattedBookings = await Promise.all(
      bookings.map(booking => formatBookingOutput(booking))
    );

    res.json(formattedBookings);
  } catch (error) {
    console.error('Error fetching bookings:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});


router.post('/', authenticateToken, requireRoles(['user']), validateBookingInput, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.id;
    const data = req.body as BookingInput;

    // Use transaction with serializable isolation (matches Python)
    const result = await prisma.$transaction(async (tx) => {
      // Validate seat session belongs to user and is active
      const seatSession = await tx.seat_session.findFirst({
        where: {
          id: data.session_id,
          user_id: userId,
          session_end_time: {
            gt: new Date()
          }
        },
        include: {
          seat: true
        }
      });

      if (!seatSession) {
        throw new Error('Seat session does not belong to the user or has expired');
      }

      // Validate flights exist
      for (const flightId of [...data.departure_flights, ...data.return_flights]) {
        const flight = await tx.flight.findUnique({ where: { id: flightId } });
        if (!flight) {
          throw new Error(`Flight with ID ${flightId} not found`);
        }
      }

      // Validate extras exist and belong to selected flights
      for (const extra of data.extras) {
        const flightExtra = await tx.flight_extra.findUnique({
          where: { id: extra.id }
        });
        
        if (!flightExtra) {
          throw new Error(`Extra with ID ${extra.id} not found`);
        }

        const allFlightIds = [...data.departure_flights, ...data.return_flights];
        if (!allFlightIds.includes(flightExtra.flight_id)) {
          throw new Error('Extra does not belong to the selected flights');
        }
      }

      // Generate unique booking number
      let bookingNumber: string;
      let attempts = 0;
      do {
        bookingNumber = generateUniqueBookingNumber();
        const existing = await tx.booking.findFirst({
          where: { booking_number: bookingNumber }
        });
        if (!existing) break;
        attempts++;
      } while (attempts < 10);

      if (attempts >= 10) {
        throw new Error('Could not generate unique booking number');
      }

      // Create booking
      const booking = await tx.booking.create({
        data: {
          id: randomUUID(),
          user_id: userId,
          payment_confirmed: true,
          has_booking_insurance: data.has_booking_insurance,
          booking_number: bookingNumber,
          created_at: new Date()
        }
      });

      const allSeats = seatSession.seat;

      // Create departure flight bookings
      for (const flightId of data.departure_flights) {
        for (const seat of allSeats) {
          if (seat.flight_id === flightId) {
            const flight = await tx.flight.findUnique({
              where: { id: flightId }
            });
            
            const flightPrice = priceFromFlight(flight, seat.class_type);
            
            await tx.booking_departure_flight.create({
              data: {
                flight_id: flightId,
                booking_id: booking.id,
                seat_number: seat.seat_number,
                class_type: seat.class_type,
                price: flightPrice
              }
            });

            await checkAndUpdateFlightCapacity(tx, flightId);
          }
        }
      }

      // Create return flight bookings
      for (const flightId of data.return_flights) {
        for (const seat of allSeats) {
          if (seat.flight_id === flightId) {
            const flight = await tx.flight.findUnique({
              where: { id: flightId }
            });
            
            const flightPrice = priceFromFlight(flight, seat.class_type);
            
            await tx.booking_return_flight.create({
              data: {
                flight_id: flightId,
                booking_id: booking.id,
                seat_number: seat.seat_number,
                class_type: seat.class_type,
                price: flightPrice
              }
            });

            await checkAndUpdateFlightCapacity(tx, flightId);
          }
        }
      }

      // Create booking extras
      for (const extra of data.extras) {
        const extraObj = await tx.flight_extra.findUnique({
          where: { id: extra.id }
        });

        if (extraObj) {
          await tx.booking_flight_extra.create({
            data: {
              booking_id: booking.id,
              flight_id: extraObj.flight_id,
              extra_id: extraObj.id,
              extra_price: extraObj.price * extra.quantity,
              quantity: extra.quantity
            }
          });
        }
      }

      // Delete seat session (booking confirmed)
      await tx.seat_session.delete({
        where: { id: data.session_id }
      });

      return booking;
    }, {
      isolationLevel: 'Serializable'
    });

    res.status(201).json({ id: result.id });
  } catch (error: any) {
    console.error('Error creating booking:', error);
    
    if (error.message.includes('not found') || error.message.includes('does not belong')) {
      res.status(403).json({ error: error.message, code: 403 });
      return;
    }
    
    res.status(400).json({ error: error.message });
  }
});


router.get('/:booking_id', authenticateToken, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.id;
    const userRoles = req.user!.roles;
    const bookingId = req.params.booking_id;

    const booking = await prisma.booking.findUnique({
      where: { id: bookingId },
      include: {
        booking_departure_flight: {
          include: {
            flight: {
              include: {
                route: {
                  include: {
                    airport_route_departure_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airport_route_arrival_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airline: true
                  }
                },
                airline_aircraft: {
                  include: {
                    aircraft: true
                  }
                }
              }
            }
          }
        },
        booking_return_flight: {
          include: {
            flight: {
              include: {
                route: {
                  include: {
                    airport_route_departure_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airport_route_arrival_airport_idToairport: {
                      include: {
                        city: {
                          include: {
                            nation: true
                          }
                        }
                      }
                    },
                    airline: true
                  }
                },
                airline_aircraft: {
                  include: {
                    aircraft: true
                  }
                }
              }
            }
          }
        },
        booking_flight_extra: {
          include: {
            flight_extra: {
              include: {
                extra: true
              }
            }
          }
        }
      }
    });

    if (!booking) {
      res.status(404).json({ error: 'Booking not found' });
      return;
    }

    // Check permissions (matches Python logic)
    const isAdmin = userRoles.includes('admin');
    const isAirlineAdmin = userRoles.includes('airline-admin');

    if (booking.user_id !== userId && !isAdmin && !isAirlineAdmin) {
      res.status(403).json({ error: 'You do not have permission to view this booking' });
      return;
    }

    // For airline admins, check if booking is for their airline's flight
    if (isAirlineAdmin && !isAdmin) {
      const user = await prisma.user.findUnique({
        where: { id: userId },
        select: { airline_id: true }
      });

      if (!user?.airline_id) {
        res.status(403).json({ error: 'You do not have permission to view this booking' });
        return;
      }

      // Check if any flight in the booking belongs to the airline
      const hasAirlineFlights = booking.booking_departure_flight.some((df: any) => 
        df.flight.route.airline_id === user.airline_id
      ) || booking.booking_return_flight.some((rf: any) => 
        rf.flight.route.airline_id === user.airline_id
      );

      if (!hasAirlineFlights) {
        res.status(403).json({ error: 'You do not have permission to view this booking' });
        return;
      }
    }

    const formattedBooking = await formatBookingOutput(booking);
    res.json(formattedBooking);
  } catch (error) {
    console.error('Error fetching booking:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});


router.delete('/:booking_id', authenticateToken, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.id;
    const userRoles = req.user!.roles;
    const bookingId = req.params.booking_id;

    const booking = await prisma.booking.findUnique({
      where: { id: bookingId },
      include: {
        booking_departure_flight: true,
        booking_return_flight: true
      }
    });

    if (!booking) {
      res.status(404).json({ error: 'Booking not found' });
      return;
    }

    // Check permissions (matches Python logic)
    const isAdmin = userRoles.includes('admin');
    
    if (booking.user_id !== userId && !isAdmin) {
      res.status(403).json({ 
        error: 'You do not have permission to delete this booking',
        code: 403
      });
      return;
    }

    // Collect flight IDs before deletion to update capacity
    const flightIds = new Set<string>();
    booking.booking_departure_flight.forEach((df: any) => flightIds.add(df.flight_id));
    booking.booking_return_flight.forEach((rf: any) => flightIds.add(rf.flight_id));

    try {
      // Delete booking (cascade will handle related records)
      await prisma.booking.delete({
        where: { id: bookingId }
      });

      // Update flight capacities
      for (const flightId of flightIds) {
        await checkAndUpdateFlightCapacity(prisma, flightId);
      }

      res.json({ message: 'Booking deleted successfully' });
    } catch (error: any) {
      if (error.code === 'P2003') { // Foreign key constraint
        res.status(409).json({ 
          error: 'Booking cannot be deleted due to existing dependencies' 
        });
        return;
      }
      throw error;
    }
  } catch (error) {
    console.error('Error deleting booking:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Register OpenAPI paths
registry.registerPath({
  method: 'get',
  path: '/booking',
  description: 'List bookings with optional filtering',
  summary: 'List bookings',
  tags: ['Booking'],
  security: [{ bearerAuth: [] }],
  request: {
    query: bookingListQuerySchema,
  },
  responses: {
    200: {
      description: 'List of bookings',
      content: {
        'application/json': {
          schema: z.array(bookingOutputSchema),
        },
      },
    },
    400: {
      description: 'Bad Request',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
    403: {
      description: 'Forbidden',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
  },
});

registry.registerPath({
  method: 'post',
  path: '/booking',
  description: 'Create a new booking',
  summary: 'Create booking',
  tags: ['Booking'],
  security: [{ bearerAuth: [] }],
  request: {
    body: {
      content: {
        'application/json': {
          schema: bookingInputSchema,
        },
      },
    },
  },
  responses: {
    201: {
      description: 'Booking created successfully',
      content: {
        'application/json': {
          schema: z.object({ id: z.string() }),
        },
      },
    },
    400: {
      description: 'Bad Request',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
    403: {
      description: 'Forbidden',
      content: {
        'application/json': {
          schema: z.object({ error: z.string(), code: z.number().optional() }),
        },
      },
    },
  },
});

registry.registerPath({
  method: 'get',
  path: '/booking/{booking_id}',
  description: 'Fetch a booking by ID',
  summary: 'Get booking by ID',
  tags: ['Booking'],
  security: [{ bearerAuth: [] }],
  request: {
    params: z.object({
      booking_id: z.string().uuid(),
    }),
  },
  responses: {
    200: {
      description: 'Booking details',
      content: {
        'application/json': {
          schema: bookingOutputSchema,
        },
      },
    },
    403: {
      description: 'Forbidden',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
    404: {
      description: 'Not Found',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
  },
});

registry.registerPath({
  method: 'delete',
  path: '/booking/{booking_id}',
  description: 'Delete a booking by ID',
  summary: 'Delete booking',
  tags: ['Booking'],
  security: [{ bearerAuth: [] }],
  request: {
    params: z.object({
      booking_id: z.string().uuid(),
    }),
  },
  responses: {
    200: {
      description: 'Booking deleted successfully',
      content: {
        'application/json': {
          schema: z.object({ message: z.string() }),
        },
      },
    },
    403: {
      description: 'Forbidden',
      content: {
        'application/json': {
          schema: z.object({ error: z.string(), code: z.number().optional() }),
        },
      },
    },
    404: {
      description: 'Not Found',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
    409: {
      description: 'Conflict - cannot delete due to dependencies',
      content: {
        'application/json': {
          schema: z.object({ error: z.string() }),
        },
      },
    },
  },
});

export const bookingRouter = router; 
